libtenzir/builtins/formats/json.cpp

```diff
@@ -163,117 +173,137 @@ class doc_parser {
   }
 
   [[nodiscard]] auto parse_object(simdjson::ondemand::value v, auto builder,
-                                  size_t depth = 0u) -> bool {
+                                  size_t depth = 0u) -> result {
     auto obj = v.get_object();
     if (obj.error()) {
       report_parse_err(v, "object");
-      return false;
+      return result::failure_no_change;
     }
+    auto written_once = false;
     for (auto pair : obj) {
       if (pair.error()) {
         report_parse_err(v, "key value pair");
-        return false;
+        return written_once ? result::failure_with_write
+                            : result::failure_no_change;
       }
       auto maybe_key = pair.unescaped_key();
       if (maybe_key.error()) {
         report_parse_err(v, "key in an object");
-        return false;
+        return written_once ? result::failure_with_write
+                            : result::failure_no_change;
       }
       auto key = maybe_key.value_unsafe();
       auto val = pair.value();
       if (val.error()) {
         report_parse_err(val, fmt::format("object value at key `{}`", key));
-        return false;
+        return written_once ? result::failure_with_write
+                            : result::failure_no_change;
       }
-      bool value_parse_success = false;
+      auto value_parse_result = result::success;
       // this guards the base series_builder currently used by tql2 parse_json
       if constexpr (std::same_as<detail::multi_series_builder::record_generator,
                                  decltype(builder)>) {
-        value_parse_success = parse_value(
-          val.value_unsafe(), builder.unflattened_field(key), depth + 1);
+        auto did_unflatten = false;
+        value_parse_result
+          = parse_value(val.value_unsafe(),
+                        builder.unflattened_field(key, &did_unflatten),
+                        depth + 1);
+        written_once |= value_parse_result != result::failure_no_change;
+        written_once |= did_unflatten;
```

Comment:
We discussed that this is still not 100% correct because a write can happen even when the key is not unflattened.

We discussed that `parse_object` could either return a bool or it always returns that it wrote something even when it did not (one more hacky than the other).